berkeley-function-call-leaderboard/bfcl_eval/_llm_response_generation.py

```diff
@@ -73,228 +71,297 @@ def build_handler(model_name, temperature):
-            ):
-                print(
-                    f"Rate limit reached. Sleeping for 65 seconds. Retry {retry_count + 1}/{RETRY_LIMIT}"
-                )
-                time.sleep(RETRY_DELAY)
-                retry_count += 1
-            else:
-                # This is usually the case when the model getting stuck on one particular test case.
-                # For example, timeout error or FC model returning invalid JSON response.
-                # Since temperature is already set to 0.001, retrying the same test case will not help.
-                # So we continue the generation process and record the error message as the model response
-                print("-" * 100)
-                print(
-                    "❗️❗️ Error occurred during inference. Maximum reties reached for rate limit or other error. Continuing to next test case."
-                )
-                print(f"❗️❗️ Test case ID: {test_case['id']}, Error: {str(e)}")
-                traceback.print_exc(limit=10)
-                print("-" * 100)
+    try:
+        result, metadata = handler.inference(
+            deepcopy(test_case), include_input_log, exclude_state_log
+        )
+    except Exception as e:
+        # This is usually the case when the model getting stuck on one particular test case.
+        # For example, timeout error or FC model returning invalid JSON response.
+        # Since temperature is already set to 0.001, retrying the same test case will not help.
+        # So we continue the generation process and record the error message as the model response
+        error_block = (
+            "-" * 100
+            + "\n❗️❗️ Error occurred during inference. Continuing to next test case.\n"
+            + f"❗️❗️ Test case ID: {test_case['id']}, Error: {str(e)}\n"
+            + traceback.format_exc(limit=10)
+            + "-" * 100
+        )
+        print(error_block)
 
-                return {
-                    "id": test_case["id"],
-                    "result": f"Error during inference: {str(e)}",
-                    "traceback": traceback.format_exc()
-                }
+        result = f"Error during inference: {str(e)}"
+        metadata = {"traceback": traceback.format_exc()}
 
     result_to_write = {
         "id": test_case["id"],
         "result": result,
+        **metadata,
     }
 
-    result_to_write.update(metadata)
-
     return result_to_write
 
 
 def generate_results(args, model_name, test_cases_total):
-    update_mode = args.allow_overwrite
     handler = build_handler(model_name, args.temperature)
+    num_threads = args.num_threads
+
+    if isinstance(handler, OSSHandler):
+        handler: OSSHandler
+        is_oss_model = True
+        # For OSS models, if the user didn't explicitly set the number of threads,
+        # we default to 100 threads to speed up the inference.
+        if num_threads == 1:
+            num_threads = 100
```

Comment:
or os.cpu_count()

Reply:
For locally hosted models, this fixed num_threads value is also used to estimate the GPU hours and inference cost [here](https://github.com/HuanzhiMao/gorilla/blob/b0b3d19cff28d71057634776821555fea3d49c81/berkeley-function-call-leaderboard/bfcl_eval/eval_checker/eval_runner_helper.py#L212-L219).